content/docs/rfcs/59/README.md

```diff
@@ -94,36 +112,20 @@ message URLParams {
 
 - See <https://github.com/felicio/status-web/blob/825262c4f07a68501478116c7382862607a5544e/packages/status-js/src/utils/encode-url-data.compare.test.ts#L4>
 
-<!-- # Security/Privacy Considerations
-
-A standard track RFC in `stable` status MUST feature this section.
-A standard track RFC in `raw` or `draft` status SHOULD feature this section.
-Informational RFCs (in any state) may feature this section.
-If there are none, this section MUST explicitly state that fact.
-This section MAY contain additional relevant information, e.g. an explanation as to why there are no security consideration for the respective document. -->
+# Security Considerations
+- There is currently no method to verify the validity of the encoded data. This limitation is acknowledged in the current implementation.
```

Comment:
@rymnc this is not true.

See https://github.com/status-im/status-web/blob/eb5cbcdda3ea41c0413c5effdee2f54a85b840ab/packages/status-js/src/utils/sign-url-data.ts#L33-L51 for implementation, and https://github.com/vacp2p/rfc/pull/602#discussion_r1200862165 for additional context.

Reply:
Here i am talking about the metadata itself, and the fact that you trust someone to sign and send the correct metadata to you.

Reply:
thats why it says "validity of encoded data" :)

Reply:
In that case, I might have the terms mixed up. I'd say "validity" can be quite broad.

To be more specific, the integrity should be ensured and validated with what I've shared above. And authenticity, while yes a link can be send to you, but you could also obtain it from a profile in the app or from its bio on social. In which case, validating the authenticity depends on you trusting the source or relying on other mechanism like identicon ring, emoji hash or public key.

That's why I think the sentence is not true.

Also,

> This limitation is acknowledged in the current implementation.

acknowledged how and where?

Reply:
Would you say this better describes it?
```
The integrity of the encoded data is ensured by the signature, however, the authenticity of the data is not guaranteed.
```

Reply:
Just to make sure I understand, let's take a user for example, if we implemented the signature correctly, we can validate the authenticity of the data, since a signature will verify that display_name/description/color belongs to the extracted public key (and being user set, they are arbitrary pieces of data).

Maybe though I am missing something or misunderstanding the context

Reply:
as per what I could understand from the implementation, ref: https://github.com/status-im/status-web/blob/eb5cbcdda3ea41c0413c5effdee2f54a85b840ab/packages/status-js/src/utils/encode-url-data.ts#L25-L29

and subsequently looking at the test for it - https://github.com/status-im/status-web/blob/eb5cbcdda3ea41c0413c5effdee2f54a85b840ab/packages/status-js/src/utils/create-url.test.ts#L110-L121

looks like any arbitrary data can be encoded and signed with the community's private key, and there is no way to verify the truthfulness of it. Probably @felicio can answer better here.

Reply:
From what I can gather by looking at the code, if you own the private key of that community, you can set those fields to any value you would like.

Basically those are community-set fields, and they are verified against the private key/signature of the community, they are not validated against any other service/third party. 

 It's true that in theory they could communicate two different things, one over URL, and the other over waku, but both would be authoritative and their provenance would be certain (it would be the equivalent of a person saying to someone that their favorite color is blue and to someone else that their favorite color is green, which is inconsistent, but there's no doubt on the source)

But @felicio might want to weight in as I don't have as much context.

Reply:
I think though that's difficult to explain, since the data is authenticated against the key, so I think to say that the authenticity of the data is not guaranteed can be confusing. The problem of whether you can trust that particular key is a different one and has to do more on the distribution channel rather than the data itself in my opinion.

As a matter of fact, integrity of the data is tied to the public key extracted.
If the data is modified by a malicious user, the public key extraction will result in a different key, but won't throw any error .

Reply:
> authenticity of the data is not guaranteed can be confusing.

Thanks, I'll definitely take your word for it.

> The problem of whether you can trust [...]

That's what I understood to be the distinction between your guys' cases and beloved it could still be explained as different level of authenticity.

Anyway, hopefully it's clearer now.

>  but won't throw any error .

That's what I understood to be a bug, or at least in the secp256k1 implementation I've been using. Hence https://github.com/vacp2p/rfc/pull/602#discussion_r1200865413, which I'd say covers that.

Reply:
I think that's a bit of a different bug, what I mean is that in the URL, we pass

1) Data/Hash
2) PublicKey
3) Signature

All this 3 are passed over the wire and can be changed by a malicious attacker.
So verifying that extracted public key == public key is not meaningful, since all 3 of them can be changed by an attacker.
The important thing though, is that if an attacker changes anything, the extracted public key will be different, so it won't refer to the original user.

I am not sure you need the public key to be passed in the data/url, what do you use it for, since it can be extracted by the signature?

Reply:
> what do you use it for, since it can be extracted by the signature?

That's what I was referring to in the previous comment with https://github.com/vacp2p/rfc/pull/602#discussion_r1200865413.

In other words, if the key was extracted following cases from https://github.com/status-im/status-web/commit/f0e59b379817b477db68972a98c17109bd952436#diff-4c64b597af17789d552f8acb35dfc18def8c8c09dbff90f2548d91357ee272df wouldn't pass

 - `should not verify URL data and random signature`
 - `should not verify random URL data and random signature`

Reply:
I hoped to be proved wrong when work on other implementations was to be picked up, but I should have stated that explicitly anyway. So I did at https://github.com/vacp2p/rfc/pull/602#discussion_r1210569581.

Reply:
mmh, not sure I follow, I had a look at the code, what it seems to be doing is (correct me if I am wrong):

1) We have `signature`, `hash` and `publicKey`
2) We verify public key against / signature / hash
3) If that's successful, we proceed

This is ok, but I don't think it's necessary.
Verifying a signature makes sense if you have some pieces of data that you can trust (the public key from a key storage/ the hash from somewhere etc), but in this case, the only case it would fail, is if the attacker updated the content, but didn't change the public key, which doesn't seem to be a realistic case (they would update both the content & the public key).
So this is really equivalent to just recovering the public key, and in the URL scheme you would not have to pass the public key.
If an attacker modifies the content, it will recover a different key, which is expected.

To give a comparison, with the current scheme, if an attacker:

1) Modifies the payload/hash/signature, but does not modify the public key -> fail
2) Modifies the payload/hash/signature and modifies the public key key ->  validates, resulting in a different key (that's fine)

If you don't pass the public key, you have only one case, which will result in 2), which is expected.

So my instinct would be not to pass the public key in the URL scheme, as I don't think it's necessary (waku does the same, public key is extracted and not passed over)
Apologies if I am misunderstanding something

Reply:
> the code, what it seems to be doing is

That's right, in this use case. And it's different than saying"verifying that extracted public key == public key", wouldn't you say?

> Verifying a signature makes sense if you have some pieces of data that you can trust

We're verifying the signature to preserve the integrity and authenticity as discussed earlier, don't we?

And the trust is in another "degree" or "level of authenticity" as I put it or in "distribution channel" as you put it, isn't it?

> it would fail, is if the attacker updated the content, but didn't change the public key,

Or if kept the content and changed the pk.

> If an attacker modifies the content, it will recover a different key,

Reply:
> That's right, in this use case. And it's different than saying"verifying that extracted public key == public key", wouldn't you say?

Sorry, maybe I didn't explain myself well. Basically there are 2 operations that we perform, one is extracting the key from hash/signature, while the other is verification (which we use in the codebase).
Verification is essentially this:

```
func verify(hash, signature, expectedPk) {
   return extractPk(hash,signature) == expectedPk
}
```
Which is what I was referring to as non meaningful, since it makes sense if you can trust expectedPk (you retrieve it from a trusted third party), but in this case, all 3 are relayed, therefore we should assume that all 3 can be tampered with.

> We're verifying the signature to preserve the integrity and authenticity as discussed earlier, don't we?

Maybe I used the wrong words, apologies, as explained in the previous point, best to use "extract" rather than "verify", since what we are really doing is to pin the content/signature to a specific public key, but verification would require to verify against something, something trusted, which in this case is not available (we would if for example there was a ENS name, since we can trust the ENS name resolver).


> So if we can agree on that we don't mind NOT catching cases of

Just to make sure, it's not that we are not catching those cases, the only way you can catch those cases is if you have some expectations about a pk, so you are validating something against a known/trusted pk, which I don't believe is the case here (but it could also be lack of context). It would be possible to validate in some cases at sharing time (if you share a link with me on status saying that it's your link, I can check that's the case, as I know your pk).

Either of the cases above will result in a different pk being extracted, which is the only thing we can do given that we don't have a trusted source or something to verify against that is tamper proof.
So, I would be for not including the change myself, since I think it makes things a bit more opaque and it's a bit narrow in what it prevents (an attacker tampering with hash/signature but not tampering with pk). Though there might be some genuine cases that I am missing.

Reply:
> Basically there are 2 operations that we perform, one is extracting the

Currently it is WITHOUT extracting. That's what the very first comment points to.

> therefore we should assume that all 3 can be tampered with.

That may be, but under the current implementation and with the verification in place I wouldn't be able to construct a tampered url and put your PK in it without our UIs catching it and letting anybody know it's been tampered with by flagging such url or NOT displaying the content in the first place. Like in the case of our new onboarding/previewing page replacing https://join.status.im/u/0x04ac78d5b61ea49eab8849826c628e01aed9d38ef8abe2fe53444e47bec31788049a5099ebf6dbbe5bc1c39fffee1e012faf201e3775296bec0aa38a5ddcdf1df7.

> verification would require to verify against something, something trusted, which in this case is not available
> ...
> the only way you can catch those cases is if you have some expectations about a pk, so you are validating something against a known/trusted pk,

I see, and possibly mistakenly, verifying whether the url in this scenario has been tampered with and whether I ultimately can trust the PK behind it as two different things.

Reply:
Now, if we talk about an implementation WITH extracting, or recovering,

> what we are really doing is to pin the content/signature to a specific public key

...to a key that the process of extracting produces. That I think is a good way to put it.

> Either of the cases above will result in a different pk being extracted

By the way, is it common for other pk/signature algorithms to give you different pk for my signature and your tempered data instead of failing too?

> an attacker tampering with hash/signature but not tampering with pk

Ditto previous comment at https://github.com/vacp2p/rfc/pull/600#discussion_r1211662034

> So, I would be for not including the change myself

I'll revert it then 👍. Unless anybody else weighs in.

Reply:
> By the way, is it common for other pk/signature algorithms to give you different pk for my signature and your tampered data instead of failing too?


No, it's only a feature of some (not all) ECDSA schemes, as far as I know. Other schemes (RSA/DSA), do not offer recovery of the public key, so you can only verify the signature, but you would need to have a key to validate against, pulled from some kms.

Thanks for the work on the specs!